include/jemalloc/internal/rtree_inlines.h

```diff
@@ -161,11 +163,17 @@ rtree_elm_acquire(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
 
 	spin_t spinner = SPIN_INITIALIZER;
 	while (true) {
-		extent_t *extent = rtree_elm_read(elm, false);
 		/* The least significant bit serves as a lock. */
-		void *s = (void *)((uintptr_t)extent | (uintptr_t)0x1);
-		if (!atomic_cas_p(&elm->pun, (void *)extent, s)) {
-			break;
+		void *extent_and_lock = atomic_load_p(&elm->child_or_extent,
+		    ATOMIC_RELAXED);
+		if (likely(((uintptr_t)extent_and_lock & 0x1) == 0)) {
+			void *locked = (void *)((uintptr_t)extent_and_lock
+			    | (uintptr_t)0x1);
+			if (likely(atomic_compare_exchange_strong_p(
```

Comment:
I don't know the rationale for sure (ours is just copied from the C11 versions), but I have a theory:

The instructions a CAS gets compiled to (compxchg / lwarx + stwcx / ldxr + stxr) all spit out the previous value in the location in case of failure, so it would be nice to avoid an API that doesn't expose that. On x86, all this really does is save a load (and even then, I think it would be straightforward to teach the optimizer to figure out that a CAS failure followed by reload doesn't need the redundant load). Some places it could cost an extra fence though.

It lends itself to a sort of convenient idiom in a lot of places:

void stack_push(atomic_stack_node_p *head, int data) {
  stack_node *new_head = alloc_new_head(data);
  stack_node *cur_head = load(head, acquire);
  do {
    newHead->next = cur_head;
  } while (!compare_exchange_weak(head, &cur_head, new_head);
}

(this only reason I don't do that here is because of the adaptive spinning; the values we get back might be really stale by the time we wake up).

It is sort of an annoying API for stuff like this though. I think a convenience atomic_cas_foo that takes expected by value instead of pointer is a good idea.